apps/prairielearn/src/pages/studentInstanceQuestionExam/studentInstanceQuestionExam.js

```diff
@@ -93,87 +95,140 @@ function processSubmission(req, res, callback) {
   );
 }
 
-router.post('/', function (req, res, next) {
-  if (res.locals.assessment.type !== 'Exam') return next();
+router.post(
+  '/',
+  asyncHandler(async function (req, res, next) {
+    if (res.locals.assessment.type !== 'Exam') return next();
 
-  if (!res.locals.authz_result.authorized_edit) {
-    return next(error.make(403, 'Not authorized', res.locals));
-  }
-
-  if (req.body.__action === 'grade' || req.body.__action === 'save') {
-    if (res.locals.authz_result.time_limit_expired) {
-      return next(
-        error.make(403, 'time limit is expired, please go back and finish your assessment'),
-      );
-    }
-    if (req.body.__action === 'grade' && !res.locals.assessment.allow_real_time_grading) {
-      next(error.make(403, 'Real-time grading is not allowed for this assessment'));
-      return;
-    }
-    processSubmission(req, res, function (err) {
-      if (ERR(err, next)) return;
-      res.redirect(req.originalUrl);
-    });
-  } else if (req.body.__action === 'timeLimitFinish') {
-    // Only close if the timer expired due to time limit, not for access end
-    if (!res.locals.assessment_instance_time_limit_expired) {
-      return res.redirect(req.originalUrl);
+    if (!res.locals.authz_result.authorized_edit) {
+      return next(error.make(403, 'Not authorized', res.locals));
```

Comment:
If this is now an async function, you can just `throw` the error instead of explicitly passing it to `next`. Ditto elsewhere.